### ISSUE TITLE
[wm-actions] full-screened views are not set on top

### ISSUE THREAD

**Issue by travankor** [WayfireWM/wayfire#940]
**Describe the bug**
After full-screening a view, a secondary view is sometimes visible on top of the full-screened view.

**To Reproduce**
1. Enable wm-actions plugin
2. Toggle `always_on_top` view with wm-actions plugin
3. Focus a different view
4. Fullscreen view

**Expected behavior**
A newly full-screened view should always be on top.

This bug is similar to #930, and focus-follows (with `raise_on_top = false`) is also affected by the issue.

**Wayfire version**
0.6.0

**Comment by travankor** [comment]
Another indication that this is a bug is that full-screened views normally appear over layer shell views with the TOP layer. However, this isn't the case when a secondary view `always_on_top` set is present (ie: the layer shell TOP layer view is on top of the full-screened view, which is wrong).

**Comment by travankor** [comment]
Interestingly, a view with `always_on_top` set appears below layer shell TOP layer views if the view is *not* full-screened.

I'm not sure if this is a bug though (I actually prefer this behavior).

**Comment by damianatorrpm** [comment]
If I understand correctly I disagree, if I explicitly set a view as always on top, it should be above fullscreen views IMHO.

**Comment by ammen99** [comment]
FWIW in gnome, always-on-top views stay on top of a fullscreen view (as is the current behavior in Wayfire). And I think this is indeed the wanted behavior.

**Comment by travankor** [comment]
Well, the issue with focus-follows and fullscreen is still there.

Reproduce:
1. Enable wm-actions and focus-follows plugins
2. Create two views
3. Focus a different view with mouse
4. Fullscreen view

This behavior is different than if you click the second view.

 And there is still inconsistent behavior wrt layer shell surfaces that use the TOP layer. always-on-top is sometimes above them, sometimes below

**Comment by ammen99** [comment]
> This behavior is different than if you click the second view in step 3.

This specific part can be fixed easily. The view could be `bring_to_front`-ed when it is fullscreened.

> always-on-top is sometimes above them, sometimes below

How so? always-on-top is actually always below TOP, fullscreen might be higher than TOP (note: fullscreen != always-on-top)

**Comment by travankor** [comment]
>How so? always-on-top is actually always below TOP, fullscreen might be higher than TOP (note: fullscreen != always-on-top)

Full-screened views normally appear over layer shell views with the TOP layer. However, this isn't the case when a secondary view with always-on-top set is present (ie: the layer shell TOP layer view is on top of the full-screened view, which is wrong).

And this second example is also affected by focus-follows (with `raise_on_top = false`).

So, the current behavior is always-on-top<TOP\<fullscreen but always-on-top>fullscreen. Isn't this confusing and contradictory? Shouldn't always-on-top be below fullscreen? GNOME doesn't have layer-shell, so I'm not sure we should follow their behavior.

**Comment by damianatorrpm** [comment]
I don't think that is the case. There should be no panels when you're in fullscreen, so they hide.
Panels would also hide if you're having a fullscreen on top view.

I for one expect the current behaviour

**Comment by travankor** [comment]
>I don't think that is the case. There should be no panels when you're in fullscreen, so they hide.
Panels would also hide if you're having a fullscreen on top view

Sorry, but I don't understand this.The layer shell protocol says nothing about auto-hiding afaict. This must be something wayfire-specific. You can test all of this out with wofi, which is mentioned in `wayfire.ini` and defaults to TOP (although this is configureable).

>I for one expect the current behaviour

Which current behaviour? always-on-top>fullscreen? If so, can we make TOP\<fullscreen<always-on-top?

**Comment by damianatorrpm** [comment]
>Sorry, but I don't understand this.The layer shell protocol says nothing about auto-hiding afaict. This must be something wayfire->specific. You can test all of this out with wofi, which is mentioned in wayfire.ini and defaults to TOP (although this is >configureable).

Yes it's wayfire specific and I love it, panels should not be visible in fullscreen games.

>Which current behaviour? always-on-top>fullscreen? If so, can we make TOP<fullscreen<always-on-top?

Always on top is on top of fullscreen is the current behavior.
Always on top (a state of window layer) should be on top of fullscreen (a state of a window).

**Comment by ammen99** [comment]
> If so, can we make TOP<fullscreen<always-on-top?

This will require more changes to the APIs, and also make things more confusing. If I have a fullscreen window X, I use alt-tab and focus a different window Y, we have X < Y < panel (this is purposefully so, I know some other WMs don't allow alt-tab out of fullscreen). If we have an always-on-top Z, things look like X < Z < panel, and you say it should be panel < X < Z which is inconsistent with the other behavior.

**Comment by travankor** [comment]
@ammen99 

What about:

>This specific part can be fixed easily. The view could be bring_to_front-ed when it is fullscreened.

And this second part:

Reproduce:

1. Enable wm-actions plugins
2. Create view
3. Fullscreen view
4. Create view
5. Create "panel" layer view

The panel view is over full-screen view.

**Comment by travankor** [comment]
>On a second thought, why shouldn't the view raise itself when you haven't interacted with it?

Isn't focusing the view, a type of interaction? The view is not bring_to_front-ed because of the focus-follows plugin (with `raise_on_top=false`).

>This is expected, if you alt-tab out of a fullscreen view (creating a view is the same, it gets stacked on top of the fullscreen view) the panel will be visible. I.e the "fullscreen mode" is only active if the top window is really fullscreen, otherwise, you'll see panels & co

Doesn't this violate the layer-shell protocol? Or can the protocol be changed to be more flexible for this?

**Comment by ammen99** [comment]
> Doesn't this violate the layer-shell protocol? Or can the protocol be changed to be more flexible for this? And also, although switcher (alt-tab) works in this case, fast-switcher does not (specifically, the reproduction steps above).

What exactly are we violating here?

**Comment by ammen99** [comment]
> Isn't focusing the view, a type of interaction?

Yes, but I assumed you focus the view and then fullscreen it by for ex. clicking somewhere?

**Comment by travankor** [comment]
>Yes, but I assumed you focus the view and then fullscreen it by for ex. clicking somewhere?

No? By fullscreening it, shouldn't the view automatically have a higher z-order than normal (no always-on-top or similar)  views?

**Comment by travankor** [comment]
>What exactly are we violating here?

 For the top and overlay layers, the seat will always give
  exclusive keyboard focus to the top-most layer which has keyboard
  interactivity set to exclusive

**Comment by travankor** [comment]
>No? By fullscreening it, shouldn't the view automatically have a higher z-order than normal (no always-on-top or similar) views?

Specifically, here is the reproduction.

Reproduce:

1. Enable wm-actions and focus-follows (with raise_on_top = false) plugins
2. Create two views
3. Focus a different view with mouse
4. Fullscreen view

**Comment by ammen99** [comment]
> > 
> 
> No? By fullscreening it, shouldn't the view automatically have a higher z-order than normal (no always-on-top or similar) views?

Where is that specified? Wayfire specifically allows you to have fullscreen views everywhere, not just on top.

**Comment by travankor** [comment]
This is what KDE and sway do. I would imagine GNOME does something similar?

>Wayfire specifically allows you to have fullscreen views everywhere, not just on top.

Right, but isn't this about setting the default behavior? You could use always_on_bottom + fullscreen to accomplish alternative fullscreens.
